text/trace/0168-sampling-propagation.md

```diff
@@ -0,0 +1,181 @@
+# Propagate head trace sampling probability
+
+Use the W3C trace context to convey consistent head trace sampling probability.
+
+## Motivation
+
+The head trace sampling probability is the probability associated with
+the start of a trace context that determines whether child contexts
+are sampled or not when using the `ParentBased` Sampler.  It is useful
+to know the head trace sampling probability associated with a context
+in order to build span-to-metrics pipelines when the built-in
+`ParentBased` Sampler is used.
+
+A consistent trace sampling decision is one that can be carried out at
+any node in a trace, which supports collecting partial traces.
+OpenTelemetry specifies a built-in `TraceIDRatioBased` Sampler that
+aims to accomplish this goal but was left incomplete (see
+[TODOs](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#traceidratiobased) in the specification).
+
+We propose to propagate the necessary information alongside the [W3C
+sampled flag](https://www.w3.org/TR/trace-context/#sampled-flag) using
+`tracestate` with an `otelprob` vendor tag.
+
+## Explanation
+
+Two pieces of information are needed to convey consistent head trace
+sampling probability:
+
+1. The head trace sampling probability
+2. Source of consistent sampling decisions.
+
+This proposal uses one byte of information for each of these.
+
+### Probability value
+
+To limit the cost of this extension and for statistical reasons
+documented below, we propose to limit head trace sampling probability
+to powers of two.  This limits the available head trace sampling
+probabilities to 1/2, 1/4, 1/8, and so on.  We can compactly encode
+these probabilities as small integer values using the base-2 logarithm
+of the adjusted count (i.e., inverse probability).
+
+For example, the probability value 2 corresponds with 1-in-4 sampling,
+the probability value 10 corresponds with 1-in-1024 sampling.  Using
+one byte of information we can convey sampling rates as small as 2^-255.
+
+### Random value
+
+With head trace sampling probabilities limited to powers of two, the
+amount of randomness needed per trace context is limited.  A
+consistent sampling decision is accomplished by propagating a
+geometrically distributed random variable with shape parameter `1/2`,
+requiring only two bits of randomness on average per trace.  See
+[Estimation from Partially Sampled Distributed
+Traces](https://arxiv.org/pdf/2107.07703.pdf) section 2.8 for a
+detailed explanation.
+
+Such a random variable `r` can be generated using the following
+pseudocode:
+
+```
+r := 0
+for {
+  if nextRandomBit() {
+    break // The expected value of r is 2
+  }
+  r++
+}
+```
+
+This can be computed from a stream of random bits as the number of
```

Comment:
if the intent is to use traceIdRatio sampler, this section is completely ignoring the randomness already present in the trace id

Reply:
Yes. This has been discussed in past meetings of the Sampling SIG, which led to this issue: https://github.com/open-telemetry/opentelemetry-specification/issues/1826

The OTel specification says very little about how random a TraceID should be. @carlosalberto wrote:

> One detail to be aware of here is that some vendors may be providing their existing ids, so we have to make sure this works in all scenarios.

if we are not able to dictate the degree of actual randomness, we may not be able to use the TraceID for effective sampling by itself. The other options that have been discussed are this option (in this PR, to attach a little additional randomness) and to specify a standard hashing function to use. Of those options, this option seems easiest and most reliable. 

Just saying "16 random bytes" doesn't say enough. If we say that every bit of the 128-bit TraceID is independently and identically random with p=0.5, then the random property of that ID we are trying to compute is the number of leading zeros. If we can assert that every bit is perfectly random, then we do not need this additional random variable to be propagated.

Reply:
I've written this OTEP under the assumption that we can assume nothing definite about TraceID randomness at the individual bit level. I believe it would take a W3C revision to do better than this. Although I am willing to continue updating this document with more possibilities if necessary, I think that's out of scope here.

Reply:
I've added to the caveats section this note about W3C and traceID randomness.